src/Components/HatComponent.cs

```diff
@@ -1,102 +1,80 @@
-                SetPosition(crownX.Value, crownY.Value);
-            UpdatePosition();
-        }
-
-        public override void Update()
-        {
-            base.Update();
-
-            if (CelesteNetSupport.loaded && Entity.GetType().FullName == "Celeste.Mod.CelesteNet.Client.Entities.Ghost")
-            {
-                UpdateForGhost();
-            }
-            else
-            {
-                Visible = HatelineModule.Instance.ShouldShowHat;
-
-                if (!HatelineModule.Instance.ShouldShowHat)
-                    return;
-
-                SetPosition(HatelineModule.Instance.CurrentX, HatelineModule.Instance.CurrentY);
-                UpdatePosition();
-            }
-
-            if (Entity == null || playerSprite == null || playerHair == null)
-                return;
-
-            FlipX = !HatelineModule.Settings.DontFlip && playerHair.Facing == Facings.Left;
-            FlipY = GravityHelperImports.IsActorInverted?.Invoke(Entity as Actor) ?? false;
-            Visible = playerSprite.CurrentAnimationID != "dreamDashIn" && playerSprite.CurrentAnimationID != "dreamDashLoop";
-        }
+[Tracked(true)]
+public class HatComponent : Sprite
+{
+    public string CrownSprite;
 
-        public void UpdateForGhost()
-        {
-            Ghost ghost = Entity as Ghost;
-            DataPlayerHat hatData = null;
+    public int CrownX, CrownY;
 
-            CelesteNetSupport.CNetComponent?.Client?.Data?.TryGetBoundRef(ghost.PlayerInfo, out hatData);
-            if (ghost == null || hatData == null)
-                return;
+    public PlayerHair playerHair => Entity.Get<PlayerHair>();
+    
+    public PlayerSprite playerSprite => Entity.Get<PlayerSprite>();
 
-            CreateHat(hatData.SelectedHat);
-            SetPosition(hatData.CrownX, hatData.CrownY);
-            UpdatePosition();
-        }
+    public HatComponent(string hatSprite = HatelineModule.HAT_NONE, int? crownX = null, int? crownY = null) : base(null, null)
+    {
+        CreateHat(hatSprite, true);
+        if (crownX != null && crownY != null)
+            SetPosition(crownX.Value, crownY.Value);
+    }
+    
+    public override void Render()
+    {
+        if (Entity == null || playerHair == null) return;
+        
+        int facing = (int)playerHair.Facing;
+        int gravityFlipMultiplier = (GravityHelperImports.IsActorInverted?.Invoke(Entity as Actor) ?? false) ? -1 : 1;
+        // The new position system defaults to the hair position instead of the player position which is 7 pixels lower.
+        Vector2 backwardsCompatibilityMovement = new Vector2(0, 7) * gravityFlipMultiplier;
+        
+        Scale = playerHair.PublicGetHairScale(0);
+        Scale.Y = Math.Abs(Scale.Y) * gravityFlipMultiplier;
+        // There is no guarantee that a skin's bangs02.png is the flipped hair, but it stands true for most skins. 
+        FlipX = playerHair.Sprite.HairFrame == 2 && GetAttribute("flip") == "true";
```

Comment:
What's the reason for using HairFrame instead of playerHair.Facing? Having never seen something like this, it seems quite strange to me. 
I guess PublicGetHairScale(0) already multiplies the Facing into Scale.X for index==0 which explains the flip attribute logic.

[... half an hour later ...]

Okay, and HairFrame and PlayerAnimMetadata.Frame values for bangs 0-2 on animations was a rabbit hole I was not at all prepared to try and figure out by looking around in ILSpy and Sprites.xml, whew 😵 

So the hair frame metadata of the current player texture frame being == 2 means that Madeline is looking the other direction due to her animation, without changing her Facing at all. Interesting and _very_ complicated to figure out. :'D

And the hat's own "Facing" type flipping just sorta falls out of PublicGetHairScale(0) but then can get reset by the "flip" == "false" logic further down. Hm.

Reply:
> So the hair frame metadata of the current player texture frame being == 2 means that Madeline is looking the other direction due to her animation, without changing her Facing at all.

Yep, exactly like that. HairFrame 2 is Madeline's hair looking backwards, while madeline faces forwards. Since the difference is in the texture itself it's hard to detect it directly.

All of your flipping deductions are correct too.

Reply:
> Yep, exactly like that. HairFrame 2 is used when Madeline (and her hair) looks backwards, while Madeline faces forwards. Since the difference is in the texture itself it's hard to detect it directly.

I guess you could add this as a comment to the code but either way I'll mark this one as resolved since it was mostly me trying to figure out what's going on

Reply:
i do think there should be a more detailed comment here just for any future readers of the code, whether it's someone trying to contribute or not. always nice to have the extra info